### ISSUE TITLE
How do I control the order in which the parameters show up?

### ISSUE THREAD

**Issue by TheGuyWhoo** [Wasted-Audio/hvcc#57]
If I wanted to arrange my parameters in a specific order, is there an argument I can pass or something? Or are they ordered in the order that the object is created?

**Comment by dromer** [comment]
Which order do you mean? is this with a specific generator?

**Comment by TheGuyWhoo** [comment]
Yes with dpf!

The order of sliders.

So say by default it's:

Pitch
detune
choas

and I want it to show up as

detune 
chaos
pitch

in the daw.

**Comment by dromer** [comment]
Hmmm, so this is determined by the order in which the parameters are created, which we simply get from the list of `receivers` passed on to the generator.

Not sure what the easiest/simplest method would be to achieve such an order.
We could add a new object to the meta.json that sets these and then enhance https://github.com/Wasted-Audio/hvcc/blob/develop/hvcc/generators/c2dpf/c2dpf.py#L32 to order the list.

Rather then having to add new creation flags to each receiver to define this, that would make things even more messy.

**Comment by GModal** [comment]
First, thank you for updating heavy to work with Python3!!

I've been playing with the source, and changed HeavyDPF.cpp from:
```
    case param{{v.display}}:
        parameter.name = "{{v.display.replace('_', ' ')}}";
        parameter.symbol = "{{v.display|lower}}";
        parameter.hints = kParameterIsAutomatable
```

To the following:

```
      case param{{v.display}}:
	parameter.name = "{{v.display.replace('__1', ':')}}";
        parameter.name = parameter.name.replace('_', ' ');
        parameter.symbol = "{{v.display.replace('__1', '_')}}";
        parameter.symbol = parameter.symbol.toLower();
        parameter.hints = kParameterIsAutomatable
```

A receive object in a Pd patch of
` r A__1_gain1 @hv_param 0 3 1`

Converts to a LV2 parameter tag of **A: gain1** , and a symbol of **a__gain1**, which doesn't seem to cause problems. 

Parameters can be arranged in blocks (A:, B:, Sec1:, Sec_C, etc.). They are still alphabetized, however. This is definitely a hack (and I don't know Python), but it does work.

The code just replaces all instances of "__1" with a colon (":") in the parameter name, then replaces that string with a single underscore for the lv2 symbol. Other characters that are not allowable in Pd symbols can be inserted in a like fashion ("__2", "__3", etc.).

**Comment by dromer** [comment]
@GModal bit of a hack indeed. I don't really like creating in-line code that modifies the DPF parameters at runtime.

Maybe we can create a python function that can be injected into jinja2 instead and does all necessary replacements.
Anyway "injecting" a `:` into the name when you don't have any such sections would look really ugly. If you simply want to force an order to your parameters without such a section every thing would get this prepended, which would look pretty weird.

Perhaps such a suggested function could take care of any such parsing as well, but now we really need to come up with some generalized convention of how to deal with this.

In principal I think doing something like prepending with `__<int>` to force an order could be useful. Although this likely won't work if you go to numbers above 9 .. so we need to come up with something better.

**Comment by GModal** [comment]
Yeah, it's definitely a hack :)... This is worse: https://github.com/GModal/ChaffVerb (in the hvcc_namefix dir)
But the hacked version isn't the default Pd patch, it's just an option...

I'm aware this approach was never a viable solution:

	- Only works while the DPF generator isn't updated
	- Works ONLY on with the DPF generator (BIG fail)
	- possibility of a parameter.symbol collision 
		- slight, and Pd itself wouldn't care about multiple identical receive objects
		- hvcc seems to anticipate symbol collisions, and generates it's own symbols in that case
	
Plus, as a transparent way to reorder the parameters (without a prefix), this won't work.

I'm not sure how a ":" would be injected unless the author added it. Reordering with just a prefixed letter (or word) doesn't require a separator (or any changes to the current code). The colon (or dash, etc.) is just for readability.

One idea: a substitution/ordering list in the <patch>.json file, something like:

{"receivers_subs": [
  {"display" : "Gain_1" , "new" : "Gain, Preamp", "order" : 1},
  {"display" : "Gain_2" , "new: : "Gain, Post", "order : 2}
] }

...where "Gain_1" and "Gain_2" are the receive obj arguments in the Pd patch.

Re: the inline codes, a 4 digit `__<int>` code shouldn't be too difficult (regex?), if that's seen as a viable way to reorder.

But there REALLY needs to be a way to use other characters in the display name. The alphanumerics that hvcc currently allows aren't sufficient. I wonder why hvcc won't even accept a dash in receive objs args, when Pure Data is OK with it. 

Reordering would be great, but at least there are work-arounds, and alphabetical is better than random. Not wild about using object creation order in the Pd file, either. Pure Data is so arcane when it comes to object order. Yes, you can do it, but it's a lot of cutting an pasting. And no way to verify order without compiling the project. Some other explicit ordering schema would be preferable.

Thank you again for the work updating hvcc!!

**Comment by dromer** [comment]
Using the meta.json to enhance this for DPF was one of my suggestions as well. But this basically would mean even more custom implementation just for DPF. It would be great if this could be applied to other generators (say javascript or unity) as well.

I'm getting into the same bottle-neck now that I'm finally getting to making some plugins of my own. Of course you can edit the source files after things have been generated, but that's quite tedious. Being able to more easily group and order parameters would certainly make things easier. Particularly since we don't have any way to create GUI code, generic host-UI without any kind of directer order is quite unusable especially if you have more than even half a dozen parameters.

Lets brainstorm more about how we can achieve this over the coming weeks. I'm also open to look at different prototypes if anyone cares to jump in on this.

**Comment by GModal** [comment]
Yes, I'm not too proficient in python, but I'd like to give the code a more in-depth look.

I know the (project).heavy.ir.json file includes the complete receivers list, in alphabetical order (with my hacked names, if I'm using them). No sure if it's written to multiple times during compiling.

**Comment by dromer** [comment]
So I noticed this bit of code today that is slightly related. It's about prioritizing execution order:
https://github.com/Wasted-Audio/hvcc/blob/develop/hvcc/interpreters/pd2hv/PdReceiveObject.py#L51-L58

Which are subsequently sorted later:
https://github.com/Wasted-Audio/hvcc/blob/develop/hvcc/core/hv2ir/HeavyGraph.py#L768-L769

As the first comment says we currently cannot use this priority for externed receivers (aka "parameters"), but maybe the syntax can be adjusted to allow for this.

For instance by appending `@hv_param:<priority>` and using that.

Bit of a brain-fart, but I'll see if this can be used at all or if it would completely break things.

**Comment by dromer** [comment]
Ok nevermind. After a quick trial it seems this is not the path to take for sorting parameters.

Will have to re-think this once more ..

**Comment by GModal** [comment]
> The enumerators do something very different, I'd definitely not put it in that section ..
Right, but this is was the easiest way to demonstrate it.

> But this only solves it for the generated UI, which is anyway pretty easy to modify. The issue was with the parameter order on the host side, which is still present with this modification.
I assume the other templates also iterate receivers + sends, though. I haven't delved deep enough to get to the source of the parameter lists. It's almost certainly not in the generators :-)

**Comment by dromer** [comment]
I'd do the sorting _before_ it goes to the templates.

But creating this list and making sure the parameter names are identical to what is used internally is probably the trickiest part.

**Comment by GModal** [comment]
Hmmm. Gonna look at meta.py first, before anything.

The generated UI might be easy to modify, but it's tedious to do so often.

**Comment by dromer** [comment]
> I see it is quite easy to add fields in meta.py.

Probably best to make this a generic meta field, so this can be reused for other generators.

**Comment by lorenzosu** [comment]
I'm also interested in this. My use case is quite simple: have some LV2 plugins on Linux of pd patches and effects I like. I know I could use plugdata but also having directly plugins to use e.g. in ardour, Carla etc. is handy. Generic GUIs are fine.
But I did notice that the parameters are always sorted in alphabetical order.

I naively thought one might re-sort them via the LV2 ttl file, but this doesn't work because the order is also defined in various source files?

**Comment by GModal** [comment]
> > I see it is quite easy to add fields in meta.py.
> 
> Probably best to make this a generic meta field, so this can be reused for other generators.

I have been using the following in meta.py, for any additional JSON configuration:

```
InnerDict = Dict[str, Union[str, float, int, bool]]     # added
InnerValue = Union[InnerDict, List[str], List[str]]     # added
```

```
class DPF(BaseModel):
              ------------
    generic:        Optional[Dict[str, InnerValue]] = None  # added
    layout:         Optional[Dict[str, InnerValue]] = None  # added
```

It's a fairly flexible JSON model. Yep, it's DPF only, and I know that's not direction you'd like to go. But it works well DPF.

And the UI template Jinja2 gets fairly messy if UI items are added there. Some things (like adding a ImGui::Button or a checkbox) are fairly straight-forward, though.

Just adding this note to show I'm still giving this thought, as there seems to be interest...although these features are probably too specific to add to hvcc...

**Comment by lorenzosu** [comment]
Do you have a simple example of how the modifications to meta.py (which I assume is hvcc/types/meta.py?) could then be used to sort parameters. I am mostly interested in LV2 DPF at the moment so that would work.

Why is parameter sorting considered too specific? At the moment it seems alphabetical order is user, which doesn't seem to make much sense.

Ideally, given one use case for hvcc is to allow people to design plugins in Pd patches and 'optimise' them by generating plugins etc.  there should be a way to do this directly within pure data? Why not add for example an (optional) paramter to the [r] message like `@order 1`? Something like (assuming 0-index):

```
[r freq @hv_param 100 2000 500 @order 1]
[r gain @hv_param 0 1 0.8 @order 0]
```

Another possible semantic (which works with Pd) would be to prepend the name with square brackets expressing order, taken from Faust (https://faustdoc.grame.fr/manual/syntax/#ordering-ui-elements). This one is nice because you would immediately see the order at the beginning and is maybe a bit more maintainable than the one above (also assuming 0-index)
```
[r [0]freq @hv_param 100 2000 500]
[r [1]gain @hv_param 0 1 0.8]
```

Finally, a possibility which is more "Pure Data aware" but more subtle unless well documented, and assuming people read the docs, would be to use right to left order. While this is not really that relevant here (because we're using [r] objects) it is usually very familiar to people using Pd.
So for example in the example shown in the documentation (https://raw.githubusercontent.com/Wasted-Audio/hvcc/develop/docs/img/docs_exposed_params.png) the gain parameter would appear before the freq one because it is first in right to left order.

Just my two cents.

**Comment by lorenzosu** [comment]
Thanks for the answer!

> How do you mean? from the perspective of the parser this is by far the easiest to implement. So it makes total sense that what comes out of the Heavy Compiler is a sorted listed of externalized parameters.

I was meaning from a user point of view, that alphabetical sorting is not the one typically used.

And I also understand the need to generalise it beyond DPF.

Meanwhile I'd still be interested in the other suggested way, but I'm not really sure what files I should look into.

**Comment by dromer** [comment]
@lorenzosu I investigated your second suggestion a bit. I think it looks elegant and it's nice to learn from Faust in this. The main thing I ran into is that the full string is used to create an internal hash.

This PR is a small change but seems to solve it for externed receive parameters.
https://github.com/Wasted-Audio/hvcc/pull/286

We still need to investigate if there are any consequences for allowing non-alphanumeric characters on the input and also if we can do the same for externed sends.

Another issue is that you need to take into account the modified name if you want to go back to a hash, ie. via `getHashForString()` function.

**Comment by lorenzosu** [comment]
@dromer thanks for looking into this.

I can't seem to test the PR, meaning that I get some make errors if I change those two files in the PR as it seems the square brackets are kept in the name?

**Comment by dromer** [comment]
@lorenzosu I've made some tweaks to try and get the ordering to work. The brackets are now dropped for the internal case as well, and in some other steps the receivers were sorted as well, which is now disabled.

Still very WIP but a quick test seems like it does work. It comes with various caveats though, like you need to take care to not use the same name twice ie. `[0]something` and `[1]something`. As they will be reduced to the same name. (we can maybe generate errors when this happens)

**Comment by dromer** [comment]
Updated this to be a bit more robust.

There is now a unit test that checks for valid receiver names.
It will now also give an error if you have conflicting names as I mentioned.

Still need to do a proper end-to-end test to see if it does what we want (and really hope it doesn't break anything ..).

**Comment by dromer** [comment]
I did a basic test for both DPF and JS generators and it seems to work alright.

This is the test-patch I used:

```
#X obj 30 36 r [0]Dingen @hv_param;
#X obj 30 252 print;
#X msg 30 175 dingen \$1;
#X obj 207 36 r [1]Bla @hv_param;
#X msg 207 175 bla \$1;
#X obj 364 36 r [3]Anders @hv_param;
#X msg 364 175 anders \$1;
#X obj 30 78 s [123]dingen;
#X obj 30 137 r [123]dingen;
#X connect 0 0 7 0;
#X connect 2 0 1 0;
#X connect 3 0 4 0;
#X connect 4 0 1 0;
#X connect 5 0 6 0;
#X connect 6 0 1 0;
#X connect 8 0 2 0;
```

Can anyone else confirm that this works for them?
You can find a standalone build here that you could place into the plugdata Toolchain for instance: https://github.com/Wasted-Audio/hvcc/actions/runs/29461038189

**Comment by dromer** [comment]
Ok nope, this does not set the order yet at all. I tried the patch with the receivers created in reverse order and they are not in the order we want:

```
#X obj 46 34 r [3]Anders @hv_param;
#X msg 46 173 anders \$1;
#X obj 250 33 r [1]Bla @hv_param;
#X msg 250 172 bla \$1;
#X obj 418 30 r [0]Dingen @hv_param;
#X obj 46 251 print;
#X msg 418 169 dingen \$1;
#X obj 418 72 s [123]dingen;
#X obj 418 131 r [123]dingen;
#X connect 0 0 1 0;
#X connect 1 0 5 0;
#X connect 2 0 3 0;
#X connect 3 0 5 0;
#X connect 4 0 7 0;
#X connect 6 0 5 0;
#X connect 8 0 6 0;
```

Let me go back to the drawing board ..

**Comment by dromer** [comment]
ok it should be fixed here: https://github.com/Wasted-Audio/hvcc/actions/runs/29507051360
